test/e2e/app-dir/next-after-app-deploy/index.test.ts

```diff
@@ -0,0 +1,126 @@
+/* eslint-env jest */
+import { nextTestSetup, isNextDev, isNextDeploy } from 'e2e-utils'
+import { retry } from 'next-test-utils'
+
+const runtimes = ['nodejs', 'edge']
+
+const WAIT_BEFORE_REVALIDATING = isNextDeploy ? 10_000 : 5_000
+
+// If we want to verify that `unstable_after()` ran its callback,
+// we need it to perform some kind of side effect (because it can't affect the response).
+// In other tests, we often use logs for this, but we don't have access to those in deploy tests.
+// So instead this test relies on calling `revalidatePath` inside `unstable_after`
+// to revalidate an ISR page '/timestamp/key/[key]', and then checking if the timestamp changed --
+// if it did, we successfully ran the callback (and performed a side effect).
+
+// This test relies on revalidating a static page, so it can't work in dev mode.
+const _describe = isNextDev ? describe.skip : describe
+
+_describe.each(runtimes)('unstable_after() in %s runtime', (runtimeValue) => {
+  const { next, skipped } = nextTestSetup({
+    files: __dirname,
+    env: { WAIT_BEFORE_REVALIDATING: WAIT_BEFORE_REVALIDATING + '' },
+  })
+  const retryDuration = WAIT_BEFORE_REVALIDATING * 2
+
+  if (skipped) return
+  const pathPrefix = '/' + runtimeValue
+
+  type PageInfo = {
+    key: string
+    timestamp: number
+  }
+
+  const getTimestampPageData = async (path: string): Promise<PageInfo> => {
+    const fullPath = `/timestamp/key/${encodeURIComponent(path)}`
+    const $ = await next.render$(fullPath)
+    const dataStr = $('#page-info').text()
+    if (!dataStr) {
+      throw new Error(`No page data found for '${fullPath}'`)
+    }
+    return JSON.parse(dataStr) as PageInfo
+  }
+
+  const getInitialTimestampPageData = async (path: string) => {
+    // FIXME: this seems like a bug in PPR --
+    // despite the page being static, the first two requests both cause a render
+    // and only the second one gets cached and re-used.
+    // we work around it by doing a dummy request to get that first "uncached" request out of the way.
+    if (process.env.__NEXT_EXPERIMENTAL_PPR) {
+      await getTimestampPageData(path)
+    }
```

Comment:
this part is sketchy, would be good to figure out why it works this way (might be a bug)